File: TimeCardClientAndroid/app/src/main/java/jp/uhimania/timecardclientandroid/ui/CalendarViewModel.kt
package jp.uhimania.timecardclientandroid.ui

import androidx.annotation.StringRes
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import jp.uhimania.timecardclientandroid.R
import jp.uhimania.timecardclientandroid.TimeCardClientApplication
import jp.uhimania.timecardclientandroid.data.BreakTime
import jp.uhimania.timecardclientandroid.data.CalendarRecord
import jp.uhimania.timecardclientandroid.data.CalendarRecordRepository
import jp.uhimania.timecardclientandroid.data.DefaultCalendarRecordRepository
import jp.uhimania.timecardclientandroid.data.TimeInterval
import jp.uhimania.timecardclientandroid.data.TimeRecord
import jp.uhimania.timecardclientandroid.data.month
import jp.uhimania.timecardclientandroid.data.startOfDay
import jp.uhimania.timecardclientandroid.data.timeIntervalSince
import jp.uhimania.timecardclientandroid.data.year
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.util.Date

data class CalendarUiState(
    val date: Date = Date(),
    val isLoading: Boolean = false,
    val records: List<CalendarRecord> = listOf(),
    @param:StringRes val message: Int? = null
) {
    data class BreakTime(
        val start: Date? = null,
        val end: Date? = null,
        val elapsed: TimeInterval? = null
    )

    data class TimeRecord(
        val checkIn: Date? = null,
        val checkOut: Date? = null,
        val elapsed: TimeInterval? = null,
        val breakTimes: List<BreakTime> = listOf()
    )

    data class CalendarRecord(
        val date: Date = Date(),
        val records: List<TimeRecord> = listOf()
    )
}

class CalendarViewModel(
    private val calendarRecordRepository: CalendarRecordRepository
) : ViewModel() {
    private val _date = MutableStateFlow(Date())
    private val _isLoading = MutableStateFlow(false)
    private val _message = MutableStateFlow<Int?>(null)

    @OptIn(ExperimentalCoroutinesApi::class)
    private val _records = _date
        .flatMapLatest { calendarRecordRepository.getRecordsStream(it.year(), it.month()) }
        .onEach { _isLoading.value = false }

    val uiState = combine(_date, _isLoading, _records, _message) { date, loading, recs, msg ->
        CalendarUiState(
            date = date,
            isLoading = loading,
            records = recs.map { it.asUiState() },
            message = msg
        )
    }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = CalendarUiState(isLoading = true)
        )

    init {
        refreshRecords()
    }

    private fun refreshRecords() {
        _isLoading.value = true

        viewModelScope.launch {
            try {
                calendarRecordRepository.refreshRecords(_date.value.year(), _date.value.month())
            } catch (_: Exception) {
                _message.value = R.string.error_get_records_failed
            }
        }
    }

    fun updateDate(date: Date) {
        _date.value = date
        refreshRecords()
    }

    fun messageShown() {
        _message.value = null
    }

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[APPLICATION_KEY] as TimeCardClientApplication
                val repository = DefaultCalendarRecordRepository(app.networkDataSource, app.localDataSource)
                CalendarViewModel(repository)
            }
        }
    }
}

fun CalendarRecord.asUiState(): CalendarUiState.CalendarRecord {
    return CalendarUiState.CalendarRecord(
        date = this.date,
        records = this.records.map { it.asUiState() }
    )
}

fun TimeRecord.asUiState(): CalendarUiState.TimeRecord {
    return CalendarUiState.TimeRecord(
        checkIn = this.checkIn,
        checkOut = this.checkOut,
        elapsed = this.checkOut?.timeIntervalSince(this.checkIn?.startOfDay() ?: Date()),
        breakTimes = this.breakTimes.map { it.asUiState() }
    )
}

fun BreakTime.asUiState(): CalendarUiState.BreakTime {
    return CalendarUiState.BreakTime(
        start = this.start,
        end = this.end,
        elapsed = this.end?.timeIntervalSince(this.start?.startOfDay() ?: Date())
    )
}

File: TimeCardClientIOS/TimeCardClientIOS/UI/CalendarViewModel.swift
//
//  CalendarViewModel.swift
//  TimeCardClientIOS
//
//  Created by uhimania on 2026/01/07.
//

import Foundation

@Observable
class CalendarViewModel {
    @Observable
    class BreakTime: Identifiable {
        let start: Date?
        let end: Date?
        
        var interval: TimeInterval {
            guard let start = start else { return 0 }
            guard let end = end else { return 0 }
            return end.timeIntervalSince(Calendar.current.startOfDay(for: start))
        }
        
        init(start: Date? = nil, end: Date? = nil) {
            self.start = start
            self.end = end
        }
    }
    
    @Observable
    class TimeRecord: Identifiable {
        let checkIn: Date?
        let checkOut: Date?
        let breakTimes: [BreakTime]
        
        var interval: TimeInterval {
            guard let checkIn = checkIn else { return 0 }
            guard let checkOut = checkOut else { return 0 }
            return checkOut.timeIntervalSince(Calendar.current.startOfDay(for: checkIn))
        }
        
        init(checkIn: Date? = nil, checkOut: Date? = nil, breakTimes: [BreakTime] = []) {
            self.checkIn = checkIn
            self.checkOut = checkOut
            self.breakTimes = breakTimes
        }
    }
    
    @Observable
    class CalendarRecord: Identifiable {
        let date: Date
        let records: [TimeRecord]
        
        init(date: Date, records: [TimeRecord] = []) {
            self.date = date
            self.records = records
        }
    }
    
    @ObservationIgnored private let repository: CalendarRecordRepository
    @ObservationIgnored private var fetchTask: Task<Void, Never>?
    
    var date: Date = .now
    var records: [CalendarRecord] = []
    var loading: Bool = false
    var message: String = ""
    
    init(_ repository: CalendarRecordRepository) {
        self.repository = repository
        fetchRecords()
    }
    
    deinit {
        fetchTask?.cancel()
    }
    
    func fetchRecords() {
        self.loading = true
        
        let stream = repository.getRecordsStream(year: date.year, month: date.month)
        
        fetchTask?.cancel()
        fetchTask = Task { [weak self] in
            do {
                for try await records in stream {
                    await MainActor.run { [weak self] in
                        self?.records = records.map { $0.asViewModel() }
                        self?.loading = false
                    }
                }
            } catch {
                await MainActor.run { [weak self] in
                    self?.message = "データを取得できませんでした"
                    self?.loading = false
                }
            }
        }
    }
}

extension CalendarRecord {
    func asViewModel() -> CalendarViewModel.CalendarRecord {
        CalendarViewModel.CalendarRecord(
            date: self.date,
            records: self.records.map { $0.asViewModel() }
        )
    }
}

extension TimeRecord {
    func asViewModel() -> CalendarViewModel.TimeRecord {
        CalendarViewModel.TimeRecord(
            checkIn: self.checkIn,
            checkOut: self.checkOut,
            breakTimes: self.breakTimes.map { $0.asViewModel() }
        )
    }
}

extension TimeRecord.BreakTime {
    func asViewModel() -> CalendarViewModel.BreakTime {
        CalendarViewModel.BreakTime(
            start: self.start,
            end: self.end
        )
    }
}
